Make RwLockReadGuard covariant

r-raymond opened the pull request:
Hi, first time contributor here, if anything is not as expected, please let me know.

`RwLockReadGoard`'s type constructor is invariant. Since it behaves like a smart pointer to an immutable reference, there is no reason that it should not be covariant. Take e.g.

```
fn test_read_guard_covariance() {
    fn do_stuff<'a>(_: RwLockReadGuard<'_, &'a i32>, _: &'a i32) {}
    let j: i32 = 5;
    let lock = RwLock::new(&j);
    {
        let i = 6;
        do_stuff(lock.read().unwrap(), &i);
    }
    drop(lock);
}
```
where the compiler complains that &i doesn't live long enough. If `RwLockReadGuard` is covariant, then the above code is accepted because the lifetime can be shorter than `'a`. 

In order for `RwLockReadGuard` to be covariant, it can't contain a full reference to the `RwLock`, which can never be covariant (because it exposes a mutable reference to the underlying data structure). By reducing the data structure to the required pieces of `RwLock`, the rest falls in place.

If there is a better way to do a test that tests successful compilation, please let me know.

Fixes #80392

rust-highfive commented:
Hey! It looks like you've submitted a new PR for the library teams!

If this PR contains changes to any `rust-lang/rust` public library APIs then please comment with `r? rust-lang/libs-api @rustbot label +T-libs-api -T-libs` to request review from a libs-api team reviewer. If you're unsure where your change falls no worries, just leave it as is and the reviewer will take a look and make a decision to forward on if necessary.

Examples of `T-libs-api` changes:

* Stabilizing library features
* Introducing insta-stable changes such as new implementations of existing stable traits on existing stable types
* Introducing new or changing existing unstable library APIs (excluding permanently unstable features / features without a tracking issue)
* Changing public documentation in ways that create new stability guarantees
* Changing observable runtime behavior of library APIs

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @Mark-Simulacrum (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

m-ou-se commented:
@rfcbot merge

RustyYato commented:
I don't think this change is sound because of problems like this:
* https://github.com/rust-lang/unsafe-code-guidelines/issues/125
* https://github.com/rust-lang/rust/issues/63787

tldr: with this change `Ref` gets the `noalias` attribute (from the reference), which is unsound.

In fact, if we check the IR we see a `noalias` attribute with the new change, but not the existing code.

https://rust.godbolt.org/z/jbe6Yah6j
```rs
struct RefCell<T: ?Sized> {
    borrow_flag: core::cell::Cell<usize>,
    cell: core::cell::UnsafeCell<T>,
}

struct RefStd<'a, T> {
    cell: &'a RefCell<T>,
}

struct Ref<'a, T> {
    cell: &'a core::cell::Cell<usize>,
    value: &'a T,
}

// has noalias
#[no_mangle] fn foo(rc: &RefCell<i32>, r: Ref<'_, i32>) {}
// does not have noalias
#[no_mangle] fn bar(rc: &RefCell<i32>, r: RefStd<'_, i32>) {}
```

generated llvm-ir
```llvm-ir
; ModuleID = 'example.6cc7c99e-cgu.0'
source_filename = "example.6cc7c99e-cgu.0"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind readnone willreturn
define void @foo({ i64, i32 }* nocapture noundef align 8 dereferenceable(16) %rc, i64* nocapture noundef align 8 dereferenceable(8) %r.0, i32* noalias nocapture noundef readonly align 4 dereferenceable(4) %r.1) unnamed_addr #0 {
start:
  ret void
}

; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind readnone willreturn
define void @bar({ i64, i32 }* nocapture noundef align 8 dereferenceable(16) %rc, i64* nocapture noundef align 8 dereferenceable(16) %r) unnamed_addr #0 {
start:
  ret void
}

attributes #0 = { mustprogress nofree norecurse nosync nounwind nonlazybind readnone willreturn "probe-stack"="__rust_probestack" "target-cpu"="x86-64" }

!llvm.module.flags = !{!0, !1}

!0 = !{i32 7, !"PIC Level", i32 2}
!1 = !{i32 2, !"RtLibUseGOT", i32 1}
```

coolreader18 commented:
Separately from the soundness concern; should the `data: &T` field be first? It's a lot more common to access the data than the mutex, and so then it could compile to the same as a pointer access without having to add an offset. (Maybe?)

r-raymond commented:
Hi @RustyYato, thanks for speaking up, I was not aware of the `noalias`. Seems like the designated "fix" (replacing by `* const`) also works in this case.

```
struct RefCell<T: ?Sized> {
    borrow_flag: core::cell::Cell<usize>,
    cell: core::cell::UnsafeCell<T>,
}

struct RefStd<'a, T> {
    cell: &'a RefCell<T>,
}

struct Ref<'a, T> {
    cell: &'a core::cell::Cell<usize>,
    value: &'a T,
}

struct RefNew<'a, T> {
    value: *const T,
    cell: &'a core::cell::Cell<usize>,
}

// has noalias
#[no_mangle] fn foo(rc: &RefCell<i32>, r: Ref<'_, i32>) {}
// does not have noalias
#[no_mangle] fn bar(rc: &RefCell<i32>, r: RefStd<'_, i32>) {}
// does not have noalias
#[no_mangle] fn baz(rc: &RefCell<i32>, r: RefNew<'_, i32>) {}
```
```
; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind readnone willreturn
define void @foo({ i64, i32 }* nocapture noundef align 8 dereferenceable(16) %rc, i64* nocapture noundef align 8 dereferenceable(8) %r.0, i32* noalias nocapture noundef readonly align 4 dereferenceable(4) %r.1) unnamed_addr #0 {
start:
  ret void
}

; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind readnone willreturn
define void @bar({ i64, i32 }* nocapture noundef align 8 dereferenceable(16) %rc, i64* nocapture noundef align 8 dereferenceable(16) %r) unnamed_addr #0 {
start:
  ret void
}

; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind readnone willreturn
define void @baz({ i64, i32 }* nocapture noundef align 8 dereferenceable(16) %rc, i32* nocapture %r.0, i64* nocapture noundef align 8 dereferenceable(8) %r.1) unnamed_addr #0 {
start:
  ret void
}
```
(see https://rust.godbolt.org/z/GrKP9oT99).

Luckily `* const` preserves our covariancy, so this still solved the initial problem. Any objections to making this modification by the existing reviewers?

r-raymond commented:
Thanks for the suggestion @coolreader18. Seems like there shouldn't be a difference for most architectures:

```
use std::ops::Deref;

struct Test1<'a, T: ?Sized> {
    cell: &'a core::cell::Cell<usize>,
    data: *const T,
}
struct Test2<'a, T: ?Sized> {
    data: *const T,
    cell: &'a core::cell::Cell<usize>,
}


impl<T: ?Sized> Deref for Test1<'_, T> {
    type Target = T;

    fn deref(&self) -> &T {
        unsafe { &*self.data }
    }
}

impl<T: ?Sized> Deref for Test2<'_, T> {
    type Target = T;

    fn deref(&self) -> &T {
        unsafe { &*self.data }
    }
}

#[no_mangle] fn t1(t: &Test1<i32>) -> i32 {
    **t
}

#[no_mangle] fn t2(t: &Test2<i32>) -> i32 {
    **t
}
```
gives
```
        .text
        .intel_syntax noprefix
        .file   "example.6cc7c99e-cgu.0"
        .section        .text.t1,"ax",@progbits
        .globl  t1
        .p2align        4, 0x90
        .type   t1,@function
t1:
        mov     rax, qword ptr [rdi + 8]
        mov     eax, dword ptr [rax]
        ret
.Lfunc_end0:
        .size   t1, .Lfunc_end0-t1

        .section        .text.t2,"ax",@progbits
        .globl  t2
        .p2align        4, 0x90
        .type   t2,@function
t2:
        mov     rax, qword ptr [rdi]
        mov     eax, dword ptr [rax]
        ret
.Lfunc_end1:
        .size   t2, .Lfunc_end1-t2

        .section        ".note.GNU-stack","",@progbits
```
AFAICT the `mov` with offset is not more expensive than the `mov` without offset. However since there is no drawback to your suggestion, I'll go ahead and make the change.

BurntSushi commented:
I would feel a lot better if the unsafe usages in this PR had safety justifications. And also that the `unsafe fn new` routine documents what conditions the caller id required to uphold for safety.

RalfJung commented:
> Luckily * const preserves our covariancy, so this still solved the initial problem. Any objections to making this modification by the existing reviewers?

You probably want to use `NonNull` to preserve the niche. But other than that, seems fine to me.

r-raymond commented:
@RalfJung thanks for the pointer. Seems like `NonNull` is potentially a bigger footgun, but it makes sense to keep the niche for optimizations. Modified in [1ab5f09](https://github.com/rust-lang/rust/pull/96820/commits/1ab5f0936e844cea525cc5a899716bfe238006a2)

@BurntSushi thanks for the concerns. I have added some documentation to the `unsafe fn new`'s of ReadGuard and WriteGuard. Let me know if you consider this sufficient. I kept it brief considering that there should really be no reason to instantiate these Guards manually, so it is more of an internal documentation.

Wrt the arguments for safety, here is why they are safe.

```
@@ -536,7 +549,7 @@ fn deref_mut(&mut self) -> &mut T {
 impl<T: ?Sized> Drop for RwLockReadGuard<'_, T> {
     fn drop(&mut self) {
         unsafe {
-            self.lock.inner.read_unlock();
+            self.inner_lock.read_unlock();
         }
     }
 }
 ```
 This really doesn't change the behavior from before. However, if the precondition documented on the `new` function holds, then the current thread has read access, in which case it is safe to call `read_unlock` according to the documentation of `MoveableRwLock`.
 
 ```
 @@ -512,7 +525,7 @@ impl<T: ?Sized> Deref for RwLockReadGuard<'_, T> {
     type Target = T;

     fn deref(&self) -> &T {
-        unsafe { &*self.lock.data.get() }
+        unsafe { self.data.as_ref() }
     }
 }
 ```
 According to the documentation of `NonNull` https://doc.rust-lang.org/std/ptr/struct.NonNull.html this is safe if
* the pointer is properly aligned
* the pointer is dereferenceable
* the pointer points to an initialized instance of T
* Rust's aliasing rules are respected (wrt the lifetime returned)

Making arguments for these points basically takes us through the entire construction of the RwLock. If we accept that `&*self.lock.data.get()` is safe, then we know than the first 3 points are given. The last point is given since the lifetime is restricted by the lifetime of the RwLockReadGuard. As long as the instance of the latter is alive, we have (shared) read access to the underlying data structure.

Finally
```
@@ -468,12 +470,23 @@ fn from(t: T) -> Self {
 }

 impl<'rwlock, T: ?Sized> RwLockReadGuard<'rwlock, T> {
+    /// Create a new instance of `RwLockReadGuard<T>` from a `RwLock<T>`.
+    ///
+    /// It is safe to call this function if and only if `lock.inner.read()` (or
+    /// `lock.inner.try_read()`) has been successfully called before instantiating this object.
     unsafe fn new(lock: &'rwlock RwLock<T>) -> LockResult<RwLockReadGuard<'rwlock, T>> {
-        poison::map_result(lock.poison.borrow(), |_| RwLockReadGuard { lock })
+        poison::map_result(lock.poison.borrow(), |_| RwLockReadGuard {
+            data: NonNull::new_unchecked(lock.data.get()),
+            inner_lock: &lock.inner,
+        })
     }
 }
 ```
The concerning point is `NonNull::new_unchecked`. Since an initialized RwLock will always hold a valid instance of the underlying object, `lock.data.get()` can't be `null`.

Let me know if these arguments are sufficient for you. Happy to dive deeper into anything mentioned here. If anyone else has comments on the above, please share.

BurntSushi commented:
@r-raymond Thanks! I generally prefer to see `SAFETY:` comments on _all_ uses of `unsafe`, but I'll leave that up to @rust-lang/libs, as I'm not terribly familiar with the convention there. But your reasoning in the comment makes sense to me. :)

r-raymond commented:
@BurntSushi: Thanks for the response. I was not aware of the norm around SAFETY comments. Can you share an example (rust and safety is a bit hard to search)? Should we abort the merge and address the issue?

RalfJung commented:
If you grep for 'SAFETY:' in the library folder of this repo, you will find hundreds of examples. :)

r-raymond commented:
@BurntSushi I've added safety comments to the unsafe usage of RwLock{Read, Write}Guard.

m-ou-se commented:
r? @cuviper - Can you review this? This PR is working around #63787, similar to what you've done in #97027.

r-raymond commented on library/std/src/sync/rwlock.rs:
```diff
@@ -100,7 +101,8 @@ unsafe impl<T: ?Sized + Send + Sync> Sync for RwLock<T> {}
 #[stable(feature = "rust1", since = "1.0.0")]
 pub struct RwLockReadGuard<'a, T: ?Sized + 'a> {
-    lock: &'a RwLock<T>,
+    data: NonNull<T>,
```
I've added a comment in [8b6c05d](https://github.com/rust-lang/rust/pull/96820/commits/8b6c05dfccb2c99378743cd43fd4067485893ba6). I'll port the assertion over from #97027, but it'll take we a couple of days.

r-raymond replied:
[cb20e25](https://github.com/rust-lang/rust/pull/96820/commits/cb20e252aded0a8b66cc7164b42a16883925b4b5) adds a test to verify that no noalias attribute is added. The test fails if we replace `NonNull<T>` with `&'a T`.

bors commented:
:umbrella: The latest upstream changes (presumably #97924) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6954506044?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/cb20e252aded0a8b66cc7164b42a16883925b4b5/checks/6954506044/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

running 339 tests
ii...i.........i....i..i.................iii........ii.i.......i.................ii..... 88/339
............i..............i................i...iii........i..i......i........i.......i. 176/339
...............i..Fi...i.....ii..i.ii.............iiii........................i.i.ii.i.. 264/339
Some tests failed in compiletest suite=codegen mode=codegen host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
failures:


---- [codegen] src/test/codegen/noalias-rwlockreadguard.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-12/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/noalias-rwlockreadguard/noalias-rwlockreadguard.ll" "/checkout/src/test/codegen/noalias-rwlockreadguard.rs" "--check-prefixes" "CHECK,NONMSVC"
stdout: none
--- stderr -------------------------------
/checkout/src/test/codegen/noalias-rwlockreadguard.rs:12:15: error: CHECK-NOT: excluded string found in input
// CHECK-NOT: noalias
              ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/noalias-rwlockreadguard/noalias-rwlockreadguard.ll:293:70: note: found here
declare void @_ZN4core9panicking5panic17h1b3cb7cbf63f3b47E([0 x i8]* noalias noundef nonnull readonly align 1, i64, %"core::panic::location::Location"* noalias noundef readonly align 8 dereferenceable(24)) unnamed_addr #4


Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/noalias-rwlockreadguard/noalias-rwlockreadguard.ll
Check file: /checkout/src/test/codegen/noalias-rwlockreadguard.rs

-dump-input=help explains the following input dump.
Input was:
<<<<<<
        .
        .
        .
        .
      288: ; Function Attrs: argmemonly nofree nosync nounwind willreturn
      289: declare void @llvm.lifetime.end.p0i8(i64 immarg, i8* nocapture) #3
      291: ; core::panicking::panic
      291: ; core::panicking::panic
      292: ; Function Attrs: cold noinline noreturn nonlazybind uwtable
      293: declare void @_ZN4core9panicking5panic17h1b3cb7cbf63f3b47E([0 x i8]* noalias noundef nonnull readonly align 1, i64, %"core::panic::location::Location"* noalias noundef readonly align 8 dereferenceable(24)) unnamed_addr #4
not:12                                                                          !~~~~~~                                                                                                                                                  error: no match expected
      294: 
      295: ; std::sys::unix::locks::futex_rwlock::RwLock::wake_writer_or_readers
      296: ; Function Attrs: cold nonlazybind uwtable
      297: declare void @_ZN3std3sys4unix5locks12futex_rwlock6RwLock22wake_writer_or_readers17hfe00d6cdfff8143aE(%"std::sys::unix::locks::futex_rwlock::RwLock"* noundef align 4 dereferenceable(8), i32) unnamed_addr #5
        .
        .
        .
>>>>>>
>>>>>>
------------------------------------------



failures:
    [codegen] src/test/codegen/noalias-rwlockreadguard.rs
test result: FAILED. 282 passed; 1 failed; 56 ignored; 0 measured; 0 filtered out; finished in 3.29s

Build completed unsuccessfully in 0:11:06
```

</details>

cuviper commented on src/test/codegen/noalias-rwlockreadguard.rs:
```diff
@@ -0,0 +1,15 @@
+// CHECK-NOT: noalias
+// CHECK-SAME: %_data
+#[no_mangle]
+pub unsafe fn maybe_aliased(_: RwLockReadGuard<'_, i32>, _data: &UnsafeCell<i32>) {}
```
Can you change the second parameter to `&RwLock<i32>`? That makes a more realistic example of aliasing, whereas only the standard library itself ever has access to the underlying `UnsafeCell` that would alias a guard.

r-raymond replied:
[048a80](https://github.com/rust-lang/rust/pull/96820/commits/048a80140b82471b671e7ae17f6dce2bb604f473)

cuviper commented:
@bors r+

JohnTitor commented:
Seems failed in rollup: https://github.com/rust-lang/rust/pull/98367#issuecomment-1162686450
@bors r-

r-raymond commented:
> Seems failed in rollup: [#98367 (comment)](https://github.com/rust-lang/rust/pull/98367#issuecomment-1162686450) @bors r-

I have taken a stab at fixing this in [5656de7](https://github.com/rust-lang/rust/pull/96820/commits/5656de73d9541175e7dc20923995c54c0868ded6). However, I don't have access to a windows box. Is there any way I can verify this without one?

cuviper commented:
I don't know how to test it without Windows, but let's just try it without a rollup, so we don't break others' CI.

@bors r+ rollup=never

bors commented:
:pushpin: Commit 5656de73d9541175e7dc20923995c54c0868ded6 has been approved by `cuviper`

<!-- @bors r=cuviper 5656de73d9541175e7dc20923995c54c0868ded6 -->
<!-- homu: {"type":"Approved","sha":"5656de73d9541175e7dc20923995c54c0868ded6","approver":"cuviper"} -->

bors commented:
:evergreen_tree: The tree is currently [closed](https://github.com/rust-lang/rust/pull/98381#issuecomment-1163241440) for pull requests below priority 1000. This pull request will be tested once the tree is reopened.

bors commented:
:hourglass: Testing commit 5656de73d9541175e7dc20923995c54c0868ded6 with merge 00ce47209dfdd8ef8871c6ec804f0e0e04d10702...
<!-- homu: {"type":"BuildStarted","head_sha":"5656de73d9541175e7dc20923995c54c0868ded6","merge_sha":"00ce47209dfdd8ef8871c6ec804f0e0e04d10702"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/7054687119?check_suite_focus=true)
Approved by: cuviper
Pushing 00ce47209dfdd8ef8871c6ec804f0e0e04d10702 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cuviper","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/7054687119?check_suite_focus=true"},"merge_sha":"00ce47209dfdd8ef8871c6ec804f0e0e04d10702"} -->

rust-timer commented:
Finished benchmarking commit (00ce47209dfdd8ef8871c6ec804f0e0e04d10702): [comparison url](https://perf.rust-lang.org/compare.html?start=1aabd8a4a6e1871f14e804302bd60dfcbffd5761&end=00ce47209dfdd8ef8871c6ec804f0e0e04d10702&stat=instructions:u).


### [Instruction count](https://perf.rust-lang.org/compare.html?start=1aabd8a4a6e1871f14e804302bd60dfcbffd5761&end=00ce47209dfdd8ef8871c6ec804f0e0e04d10702&stat=instructions:u)

- Primary benchmarks: no relevant changes found
- Secondary benchmarks: 🎉 relevant improvement found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | N/A      | N/A | 0         |
| Regressions 😿 <br /> (secondary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (primary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (secondary) | -0.7%    | -0.7% | 1         |
| All 😿🎉 (primary) | N/A      | N/A | 0         |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=1aabd8a4a6e1871f14e804302bd60dfcbffd5761&end=00ce47209dfdd8ef8871c6ec804f0e0e04d10702&stat=max-rss)
<details>
<summary>Results</summary>


- Primary benchmarks: 😿 relevant regressions found
- Secondary benchmarks: 😿 relevant regressions found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | 2.6%     | 3.0% | 2         |
| Regressions 😿 <br /> (secondary) | 6.4%     | 9.1% | 2         |
| Improvements 🎉 <br /> (primary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (secondary) | N/A      | N/A | 0         |
| All 😿🎉 (primary) | 2.6%     | 3.0% | 2         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=1aabd8a4a6e1871f14e804302bd60dfcbffd5761&end=00ce47209dfdd8ef8871c6ec804f0e0e04d10702&stat=cycles:u)
<details>
<summary>Results</summary>


- Primary benchmarks: 🎉 relevant improvement found
- Secondary benchmarks: 😿 relevant regression found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | N/A      | N/A | 0         |
| Regressions 😿 <br /> (secondary) | 4.2%     | 4.2% | 1         |
| Improvements 🎉 <br /> (primary) | -2.9%    | -2.9% | 1         |
| Improvements 🎉 <br /> (secondary) | N/A      | N/A | 0         |
| All 😿🎉 (primary) | -2.9%    | -2.9% | 1         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression
